src/gas_metering/mod.rs

```diff
@@ -735,28 +772,65 @@ fn insert_metering_calls(
         .unwrap();
     for (original_pos, instr) in operators.iter().enumerate() {
         // If there the next block starts at this position, inject metering func_body.
-        let used_block = if let Some(block) = block_iter.peek() {
+        if let Some(block) = block_iter.peek() {
             if block.start_pos == original_pos {
                 new_func.instruction(&wasm_encoder::Instruction::I64Const(block.cost as i64));
                 new_func.instruction(&wasm_encoder::Instruction::Call(gas_func));
-                true
-            } else {
-                false
-            }
-        } else {
-            false
-        };
 
-        if used_block {
-            block_iter.next();
+                block_iter.next();
+            }
         }
 
+        // if this instruction requires dynamic gas charge calculation, inject that code
+        if let Some(metered_instr) = instr_iter.peek() {
+            if metered_instr.pos == original_pos {
+                // duplicate stack top
+                new_func.instruction(&wasm_encoder::Instruction::LocalTee(temp_local_idx));
+                new_func.instruction(&wasm_encoder::Instruction::LocalGet(temp_local_idx));
+                new_func.instruction(&wasm_encoder::Instruction::LocalGet(temp_local_idx));
+
+                // check that the argument is positive
+                //
+                // NOTE(negative bulk instruction arg):
+                // right now this instrumentation is mostly meant for bulk memory instructions
+                // In the formal spec instructions are NOT required to trap when the "count" argument
+                // is negative, and if the spec is followed exactly, those instructions may take
+                // very long to trap with a negative argument.
```

Comment:
What if we just treat them as unsigned? I.e., just charge a lot of gas?

Reply:
(if it's an i32, that is).

Reply:
That can get all kinds of messy with overflows, but I can dabble with that idea fol a bit

Reply:
Yeah, there's a separate instruction for sign extending (`I64ExtendI32I`).

Reply:
Actually, are you sure this _is_ signed? Wasm usually doesn't distinguish between teh two.